Prune stale HTTPRoutes when tags are removed from Ingress rules

kahirokunn opened the pull request:
- Add pruning in reconcileIngress to delete owned HTTPRoutes not referenced
- Add unit test to verify stale route deletion behavior

/kind bug

Fixes #895

Release Note:
```release-note
Fix a bug where HTTPRoutes could remain after tags were removed. Stale routes are now deleted during reconciliation.
```

kahirokunn commented:
@dprotaso Hello! I'd be grateful if you could leave a review when you have a moment! 🙏

kahirokunn commented:
@dprotaso Hello 🤚

dprotaso commented on pkg/reconciler/ingress/ingress.go:
```diff
@@ -123,6 +129,26 @@ func (c *Reconciler) reconcileIngress(ctx context.Context, ing *v1alpha1.Ingress
+			// Don't touch routes not owned by this Ingress
+			if !metav1.IsControlledBy(r, ing) {
+				continue
+			}
```
With the label mentioned above we can drop this

dprotaso reviewed (commented):
Change looks great - thanks for finding the bug.

Just looking for a minor tweak

kahirokunn commented on pkg/reconciler/ingress/ingress.go:
```diff
@@ -123,6 +129,26 @@ func (c *Reconciler) reconcileIngress(ctx context.Context, ing *v1alpha1.Ingress
 
+	// Delete HTTPRoutes that don't exist in the current Spec (i.e., tags removed and no longer referenced)
+	{
+		existingRoutes, err := c.httprouteLister.HTTPRoutes(ing.Namespace).List(labels.Everything())
```
Thx. fixed.

kahirokunn commented:
@dprotaso I've addressed all the points you raised. When you have a moment, I would appreciate it if you could take another look 🙏

dprotaso commented on pkg/reconciler/ingress/resources/httproute.go:
```diff
@@ -204,6 +204,7 @@ func MakeHTTPRoute(
 			Name:      LongestHost(rule.Hosts),
 			Namespace: ing.Namespace,
 			Labels: kmeta.UnionMaps(ing.Labels, map[string]string{
+				networking.IngressLabelKey:    ing.Name,
```
I wonder why our tests were passing before without this label lol

kahirokunn commented:
@dprotaso I've addressed all the points you raised. All tests have been passed. When you have a moment, I would appreciate it if you could take another look 🙏

dprotaso commented:
/lgtm
/approve
/override "codecov/patch"

dprotaso commented:
thanks @kahirokunn 🎉
